tests/test_DG_advection.py

```diff
@@ -0,0 +1,107 @@
+from dcore import *
+from firedrake import IcosahedralSphereMesh, Expression, SpatialCoordinate, \
+    as_vector, VectorFunctionSpace, File
+import itertools
+from math import pi
+
+
+def setup_DGadvection(dirname, vector=False):
+
+    refinements = 3  # number of horizontal cells = 20*(4^refinements)
+    R = 1.
+    dt = pi/3*0.001
+
+    mesh = IcosahedralSphereMesh(radius=R,
+                                 refinement_level=refinements)
+    global_normal = Expression(("x[0]/sqrt(x[0]*x[0]+x[1]*x[1]+x[2]*x[2])",
+                                "x[1]/sqrt(x[0]*x[0]+x[1]*x[1]+x[2]*x[2])",
+                                "x[2]/sqrt(x[0]*x[0]+x[1]*x[1]+x[2]*x[2])"))
+    mesh.init_cell_orientations(global_normal)
+
+    fieldlist = ['u','D']
+    timestepping = TimesteppingParameters(dt=dt)
+    output = OutputParameters(dumpfreq=150, dirname='results/tests/'+dirname)
+    parameters = ShallowWaterParameters()
+
+    state = ShallowWaterState(mesh, vertical_degree=None, horizontal_degree=2,
+                              family="BDM",
+                              timestepping=timestepping,
+                              output=output,
+                              parameters=parameters,
+                              fieldlist=fieldlist)
+
+    # interpolate initial conditions
+    u0 = Function(state.V[0], name="velocity")
+    x = SpatialCoordinate(mesh)
+    uexpr = as_vector([-x[1], x[0], 0.0])
+    u0.project(uexpr)
+
+    if vector:
+        VectorDGSpace = VectorFunctionSpace(mesh, "DG", 1)
+        f = Function(VectorDGSpace, name="f")
+        fexpr = Expression(("exp(-pow(x[2],2) - pow(x[1],2))", "0.0", "0.0"))
+        f_end = Function(VectorDGSpace)
+        f_end_expr = Expression(("exp(-pow(x[2],2) - pow(x[0],2))","0","0"))
+    else:
+        f = Function(state.V[1], name='f')
+        fexpr = Expression("exp(-pow(x[2],2) - pow(x[1],2))")
+        f_end = Function(state.V[1])
+        f_end_expr = Expression("exp(-pow(x[2],2) - pow(x[0],2))")
+
+    f.interpolate(fexpr)
+    f_end.interpolate(f_end_expr)
+
+    return state, u0, f, f_end
+
+
+def run(continuity=False, vector=False):
```

Comment:
It's somewhat unpleasant that the test writes a load of output whereever it's run.  You can get pytest to provide you with a temporary directory like so:

```python
def test_foo(tmpdir, ...):
    # tmpdir is a per-test temporary directory.
    dirname = str(tmpdir.join("DGAdvection/continuity" + str(continuity)))
```